pkg/common/passwords.go

```diff
@@ -0,0 +1,237 @@
+			fernetJSON, _ := generateFernetContent()
+			passwords[p] = fernetJSON
+		} else if p == "KeystoneCredential0" || p == "KeystoneCredential1" || p == "KeystoneFernetKey0" || p == "KeystoneFernetKey1" {
+			passwords[p] = generateFernetKey()
+		} else if p == "MigrationSshKey" {
+			migrationJSON, _ := generateSSHKeypair()
+			passwords[p] = migrationJSON
+		} else if p == "BarbicanSimpleCryptoKek" {
+			passwords[p] = generateFernetKey()
+		} else if p == "MysqlRootPassword" {
+			passwords[p] = k8s_rand.String(10)
+		} else if p == "RabbitCookie" {
+			passwords[p] = k8s_rand.String(20)
+		} else if p == "PcsdPassword" {
+			passwords[p] = k8s_rand.String(16)
+		} else if p == "HorizonSecret" {
+			passwords[p] = k8s_rand.String(10)
+		} else if p == "HeatAuthEncryptionKey" {
+			passwords[p] = k8s_rand.String(32)
+		} else if p == "OctaviaServerCertsKeyPassphrase" {
+			passwords[p] = k8s_rand.String(32)
+		} else if p == "DesignateRndcKey" {
+			passwords[p] = generateDesignateHMAC()
+		} else {
+			passwords[p] = k8s_rand.String(minPasswordSize)
+		}
+	}
+	return passwords
+}
+
+func generateCephKey() string {
+	math_rand.Seed(time.Now().UnixNano())
+
+	// https://github.com/ceph/ceph-deploy/blob/master/ceph_deploy/new.py#L21
+	key := make([]byte, 16)
+	for i := 0; i < 16; i++ {
+		key[i] = byte(math_rand.Intn(10))
+	}
+
+	header := make([]byte, 12)
+	binary.LittleEndian.PutUint16(header[0:], 1)                         // # le16 type: CEPH_CRYPTO_AES
+	binary.LittleEndian.PutUint32(header[2:], uint32(time.Now().Unix())) // le32 created: seconds
+	binary.LittleEndian.PutUint32(header[6:], 0)                         // le32 created: nanoseconds
+	binary.LittleEndian.PutUint16(header[10:], 12)                       // le16: len(key)
+
+	return base64.StdEncoding.EncodeToString(append(header, key...))
+}
+
+func generateDesignateHMAC() string {
+	key := make([]byte, 16)
+	for i := 0; i < 16; i++ {
+		key[i] = byte(math_rand.Intn(10))
+	}
+	mac := hmac.New(sha256.New, key)
+	mac.Write(key)
+	return base64.StdEncoding.EncodeToString(mac.Sum(nil))
+}
+
+func generateFernetKey() string {
+	math_rand.Seed(time.Now().UnixNano())
+	data := make([]byte, 32)
+	for i := 0; i < 32; i++ {
+		data[i] = byte(math_rand.Intn(10))
+	}
+	return base64.StdEncoding.EncodeToString(data)
+}
+
+type fernetKeyContent struct {
+	Content string `json:"content"`
+}
+
+type fernetKey struct {
+	Key1 fernetKeyContent `json:"/etc/keystone/fernet-keys/0"`
+	Key2 fernetKeyContent `json:"/etc/keystone/fernet-keys/1"`
+}
+
+func generateFernetContent() (string, error) {
+	f := fernetKey{fernetKeyContent{generateFernetKey()}, fernetKeyContent{generateFernetKey()}}
+	b, err := json.Marshal(f)
+	if err != nil {
+		return "", err
+	}
+	return string(b), err
+
+}
+
+type privateKey struct {
+	PrivateKey string `json:"private_key"`
+	PublicKey  string `json:"public_key"`
+}
+
+func generateSSHKeypair() (string, error) {
+
+	sshKey, err := rsa.GenerateKey(cryptorand.Reader, 2048)
```

Comment:
Functionally this works fine. But I just realized we actually already have some code to generate ssh keys which we should try to reuse here. No harm in landing this and follow up to add that though.

Reply:
agreed